### PR TITLE
Add support for GRPC for agent / server calls

### PR DESCRIPTION
**Pull request by Aestek**
This PR aims a starting replacing the current RPC protocol with GRPC. It starts by implementing the `Health.ServiceNodes` call as a POC.

Rationale: 
* Protobuf provides better performance and and smaller serialization size. From my very early (more on that later) benchmarks on a Health.ServiceNode result with ~1K entries, serialization is arround 4x faster and 2x smaller.
* The current RPC library is no longer maintained
* GRPC has built-in support for connection pooling and pipelining via HTTP/2

### Structs conversion
The structs `ServiceSpecificRequet` and `IndexedCheckServiceNodes` as well as the structs used by thoose two have been converted to protobuf. Gogoprotos extentions were used to produce the same Go strucsts than the original to ensure compat with the raft state, older clients etc. 
Two types cound not be mapped to protobuf, even with gogo extentions : `map[string]interface{}` and `map[string][]string`. Custom proto types have been implemented for thoose.
The only exception are the custom slice types such as `type CheckServiceNodes []CheckServiceNode` that are now plain slices, but that does not cause any issues as they marshal / unmarshal the same way.

### GRPC Server
The servers advertise that they support GRPC by providing their `grpc_port` (8310 by default) in Serf. When making a call the agent first checks that the server supports GRPC, and that the method is implemented (currently on `Health.ServiceNodes` is, and the implemented list is hardcoded on the client side).
If both conditions are true then GRPC is used, classic RPC is used.
The behavior is the same for server to server and cross DC calls.   

### TODO
- [ ] More benchmarking with various use cases
- [ ] TLS
- [ ] Retry on failure on par with RPC
- [ ] Cleanup leftover custom slice types
- [ ] Metrics

### CODE REVIEW

**Comment by pierresouchay** [comment]
@banks here is the POC I discussed with you. We plan to test it in real world (just for health to see real world impact in our largest DC) if you would consider it for inclusion eventually

**Comment by Aestek** [comment]
I'added some benchmarks, here are the results :
```
benchmark                                            old ns/op     new ns/op     delta
BenchmarkIndexedCheckserviceNodesEncoding/10-8       175084        79853         -54.39%
BenchmarkIndexedCheckserviceNodesEncoding/100-8      1718008       768297        -55.28%
BenchmarkIndexedCheckserviceNodesEncoding/500-8      8649143       3834982       -55.66%
BenchmarkIndexedCheckserviceNodesEncoding/1000-8     17097160      7758730       -54.62%

benchmark                                            old allocs     new allocs     delta
BenchmarkIndexedCheckserviceNodesEncoding/10-8       3337           547            -83.61%
BenchmarkIndexedCheckserviceNodesEncoding/100-8      33217          5407           -83.72%
BenchmarkIndexedCheckserviceNodesEncoding/500-8      166017         27009          -83.73%
BenchmarkIndexedCheckserviceNodesEncoding/1000-8     332017         54013          -83.73%

benchmark                                            old bytes     new bytes     delta
BenchmarkIndexedCheckserviceNodesEncoding/10-8       57385         28714         -49.96%
BenchmarkIndexedCheckserviceNodesEncoding/100-8      569852        281963        -50.52%
BenchmarkIndexedCheckserviceNodesEncoding/500-8      2847478       1399372       -50.86%
BenchmarkIndexedCheckserviceNodesEncoding/1000-8     5694503       2790210       -51.00%
```

**Comment by Aestek** [comment]
## Load testing
We ran some tests on preproduction, here are the results : 

We deployed the version with GRPC on one server and one client, then ran apache benchmark (concurrency 800, forever) against `/v1/health/service/consul-agent-http?stale`. This benchmark highlights the encoding differences as the queries are not blocking.
As the client switched server over time we observed the difference in load and rps (sorry for the ugly prometheus graphs) : 

Load : 
image
We can see a significant load reduction on consu03 of ~40%

RPS:
image
We can see a ~78% RPS increase 

Response time: 
This is by no mean scientific as we have no metrics for this bu by looking at the logs the response time seemed much lower (~800ms vs 1.1s). 

## Analysis
As stated before this benchmark focuses on encoding performance, however we are confident that this will result in performance improvements for Consul in the real world. 
The server running this patch ran fine along with other server and no disruption of service was experienced.

**Comment by banks** [comment]
@pierresouchay @Aestek, thanks for this work, this is really useful data. We will consider this along with our roadmap of other important performance work.

If we incorporated gRPC progressively for real I think we'd want to use the same RPC port we have now rather than need a whole new one to be opened - we already switch between 3 different protocols over that port based on the first byte. Last time I looked though there were some intricacies with getting Go's gRPC server lib to work with a connection established by an external listener so we'd have to look into that.

**Comment by Aestek** [comment]
@banks GRPC now uses the same port as classic RPC. Though it would be hacky but it was pretty straightforward.

File: agent/consul/rpc_client.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+type RPCClient struct {
+	rpcPool   *pool.ConnPool
+	grpcConns sync.Map
```
Comment (banks) [inline agent/consul/rpc_client.go]: FYI `sync.Map` is not really suited for this. I realise this is a PoC so no worries but in general in case you weren't aware map and mutex is better for anything that is not more or less "write once, read many" like global program config only updated on startup or HUP. See https://golang.org/pkg/sync/#Map. :)

File: agent/consul/rpc_client.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		return conn.(*grpc.ClientConn), nil
+	}
+
+	co, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithDialer(dialGRPC))
+	if err != nil {
+		return nil, err
+	}
+
+	c.grpcConns.Store(addr, co)
+	return co, nil
```
Comment (banks) [inline agent/consul/rpc_client.go]: I wonder if it's possible to not re-invent grpc client's connection pool here?

Comment (banks) [reply agent/consul/rpc_client.go]: It's a bit obscure but the whole client-side loadbalancing thing is what I mean: https://godoc.org/google.golang.org/grpc/balancer

you still call dial but the "conn" returned is an abstract thing that will actually get requests loadbalanced to multiple servers according to your balancer and resolver methods you pass. This way we could recreate the retry during leader election and periodic load spreading we do in the current RPC layer which would be important for long-term adoption of gRPC.

Not asking you to do this unless you want to @Aestek , just leaving a note for myself since I looked :)

Comment (Aestek) [reply agent/consul/rpc_client.go]: yes I've looked into this as well but it was easier and safer to just reuse the mechanism already in place for classic RPC. But definitely something to look into in the future.

File: agent/consul/rpc_client.go (hunk 1 above)
Comment (Aestek) [reply agent/consul/rpc_client.go]: It seemed to me that this would be a good use case as the server list seldom changes, and I was afraid of adding a lock in the hot path as this is used when forwarding to the leader on the server side.
It might be premature optimization and not make sense here. I can replace this with a classic map+lock if you prefer.

Comment (banks) [reply agent/consul/rpc_client.go]: Yeah when I wrote that I thought this would change every time a connection got dropped and recycled like a regular pool which is much more common but I think actually grpc "conns" are more abstract than that and will automatically reconnect in the background for you right? I personally still avoid `sync.Map` until a benchmark shows mutex contention but yeah it might be fine here :).

Comment (Aestek) [reply agent/consul/rpc_client.go]: Yes there should be only one GRPC conn per server with auto reconnect if needed. The actual TCP connections are handled under the hood.
Ill change this to a regular mutex for now, this is easy enough to switch back to a `sync.Map` if needed, and as you pointed out the real target is to use GRPC's load balancer.

File: agent/consul/rpc_client.go (hunk 2 above)
Comment (banks) [reply agent/consul/rpc_client.go]: I looked into this more and I agree this is a pretty complicated issue. For a start the grpc balancer interface and mechanism is pretty hard to understand and not relly documented well. The best place I found to understand it eventually was the original proposal doc for the most recent API change: https://github.com/menghanl/proposal/blob/8043d465114dfd9cba04ba96a87af6534e8a1519/L9-go-resolver-balancer-API.md

It would be possible I think to define custom resolver and balancer that replicate the current behaviour around server selection and get updated by serf etc. But it would be a pretty huge change in our router package.

Another subtlety is that the router.Manager currently uses `connPool.Ping` to check if a server is alive which is fine for now as we still have old RPCs in use and we can assume if server is up for one type of RPC it is for both, but eventually when we move to only gRPC we'd need to rethink that.

Even if we do all of that and make it work in an equivalent way, there is a bunch more thinking to put into Enterprise network area routing which comes into play for cross-DC calls between servers in different DCs.

gRPC also assumes that each RPC should be balanced between all available servers. While a custom balancer could change that, there are subtleties to consider for example typical balancer implementations initiate connections to _all_ servers and then just return the one that should be used. It's not clear to me if that means they open TCP connections to all of them eagerly or lazily but eagerly could change things a bunch for our users - maybe cause 5x more tcp connections to be opened to their servers.

gRPC also sends pings and actively managed connection status - failed connections are reported to the balancer and can be removed. It's not exactly a problem but working out how that meshes with our own Serf health for servers as handled by the Manager is not super clear to me.

As a final note, gRPC actively maintains all of the connections in the background so even if a balancer picked the same one consistently for each rebalance interval all of them would stay open. I _thought_ that was a behaviour change and that our connection pool closes unused one after rebalance but that doesn't seem to be the case from reading our code closely again so that should be OK.

tl;dr is that your approach here is a simpler way to keep all the existing logic around server selection unchanged for gRPC!

**Comment by Aestek** [comment]
## Production results
We deployed GRPC on most of our agents and one server in our biggest DC, here are the results during a deployment at ~2K rps. The GRPC version showed great improvements in CPU load and network consumption : 

### CPU usage 

As shown bellow, the GRPC version performed 40 % better than vanilla + shared blocking queries CPU-wise : 

Vanilla + SBQ : 
image

GRPC :
image

Note that in addtion to overall CPU usage, system time is greatly reduced ( / 2.5 )

Comparison : 
image

Load GRPC / Load Vanilla + SBQ : 
image

## Network

The GRPC version performed 28% better in network consumption. 

Vanilla : 
image

GRPC:
image

**Comment by banks** [comment]
Thanks @Aestek and @pierresouchay, these results are really encouraging and great data for us!

File: agent/consul/rpc_client.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	conn, ok := c.grpcConns[addr]
+	c.grpcConnsLock.RUnlock()
+	if ok {
+		return conn, nil
+	}
+
+	c.grpcConnsLock.Lock()
+	defer c.grpcConnsLock.Unlock()
+
+	co, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithDialer(dialGRPC))
```
Comment (banks) [inline agent/consul/rpc_client.go]: @Aestek do you run this over TCP without TLS currently?

Comment (Aestek) [reply agent/consul/rpc_client.go]: Yes
